### PR TITLE
Refactor kube method and add removal code

### PR DESCRIPTION
**Pull request by josephsawaya**
This commit refactors the kube method code to tear down
pods created by files that were removed in a change first,
then tearing down pods created by files that have been modified
and replacing those pods with new pods created from the contents
of the modified files.

This commit also refactors the code that gets changes and runs
the engine method based on those changes and now the change is
passed to the engineMethod func.

Fixes: #62 

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

### CODE REVIEW

File: pkg/engine/harpoon.go [hunk 1]
```diff
@@ -464,10 +416,10 @@ func (hc *HarpoonConfig) update(target *api.Target) {
...
 
 func (hc *HarpoonConfig) findDiff(gitRepo *git.Repository, directory, method, branch string) []*object.Change {
-	gitRepo, err := git.PlainOpen(directory)
-	if err != nil {
-		log.Fatalf("error while opening the repository: %s: %s", directory, err)
-	}
+	// gitRepo, err := git.PlainOpen(directory)
+	// if err != nil {
+	// 	log.Fatalf("error while opening the repository: %s: %s", directory, err)
+	// }
```
Comment (sallyom) [inline pkg/engine/harpoon.go:422]: you can remove that, my bad, thanks

File: pkg/engine/harpoon.go [hunk 2]
```diff
@@ -515,7 +467,25 @@ func (hc *HarpoonConfig) EngineMethod(ctx context.Context, path, method string,
...
+			}
+			if to != nil {
+				s, err := to.Contents()
+				if err != nil {
+					log.Fatal(err)
+				}
+				prev = &s
+			}
+		}
+		if path != "" {
```
Comment (sallyom) [inline pkg/engine/harpoon.go:485]: this will work, instead
```
if path == "" {
    return stopPods(ctx, []byte(*prev))
}
```
then, you won't need the `if path != nil` checks in kubePodman

File: pkg/engine/kube.go [hunk 3]
```diff
@@ -23,81 +24,158 @@ type YamlMeta struct {
...
 		}
-
-		b, err := k8syaml.YAMLToJSON(o)
+		err = stopPods(ctx, kubeYaml)
 		if err != nil {
 			return err
 		}
 
-		err = json.Unmarshal(b, &pod)
+		err = createPods(ctx, *path, kubeYaml)
```
Comment (sallyom) [inline pkg/engine/kube.go:53]: you can do a `return createPods(cts, *path, kubeYaml)` here, instead

**Comment by sallyom** [comment]
awesome, @josephsawaya! I left a few suggestions but this looks good!

File: pkg/engine/kube.go [hunk 4]
```diff
@@ -23,81 +24,158 @@ type YamlMeta struct {
...
 
-func kubePodman(ctx context.Context, path string) error {
-	klog.Infof("Creating podman container from %s using kube method", path)
-
-	kubeYaml, err := ioutil.ReadFile(path)
+func kubePodman(ctx context.Context, path *string, prev *string) error {
+	if path != nil {
+		klog.Infof("Creating podman container from %s using kube method", *path)
+	}
+	ctx, err := bindings.NewConnection(ctx, "unix://run/podman/podman.sock")
```
Comment (sallyom) [inline pkg/engine/kube.go:31]: ctx should be conn here, the ctx is passed in from EngineMethod

Comment (sallyom) [reply pkg/engine/kube.go:31]: oops nm, I'm wrong :) ctx is returned from NewConnection, so should be 
```
var err error
ctx, err, = bindings.NewConnection(ctx, ...)
```

Comment (sallyom) [reply pkg/engine/kube.go:31]: :shrug: I keep seeing this error: `Client not set in context`, have you seen it?  we'll look at it next week

Comment (sallyom) [reply pkg/engine/kube.go:31]: also, we have to keep in mind that if the config specifies a `directory` for TargetPath, like `examples/kube` then, if the only file in that directory is removed, upon a git push, the directory is not pushed and so no longer exists. So, when I remove example.yaml, which is the only file in the directory, I see the `no such file or directory` when looking for `examples/kube/example.yaml` but that is bc `examples/kube` no longer exists, rather than `examples/kube/example.yaml` no longer exists. I think we'll have to handle that differently?

**Review by sallyom — COMMENTED** [review]
the refactor of changes into a separate fn looks good - the remove logic we'll do more testing with, might have edge cases to figure out.

File: pkg/engine/kube.go [hunk 5]
```diff
@@ -23,81 +24,158 @@ type YamlMeta struct {
...
 		return err
 	}
-
-	for _, r := range report {
-		if _, ok := pod_map[r.Name]; ok {
-			_, err = pods.Stop(conn, r.Name, nil)
+	for _, p := range report {
+		if _, ok := pod_map[p.Name]; ok {
+			klog.Infof("Tearing down pod: %s\n", p.Name)
+			_, err = pods.Stop(ctx, p.Name, nil)
```
Comment (jeffvance) [inline pkg/engine/kube.go:81]: nit: a minor refactor could be:
```
for ...
    if _, ok := pod_map[p.Name]; ok {
    if err = stop_pod(ctx, p.Name); err != nil {
         return
    }
...

func stop_pod (ctx context.Context,  name string) error {
    klog.Infof("Tearing down pod: %s\n", name)  // **maybe want namespace here too?
    _, err := pods.Stop(ctx, name, nil)
    if err != nil {
        return err
    }
    _, err = pods.Remove(ctx, name, nil)
    if err != nil {
        return err
    }
    return nil
}
```
